File: Modules/ServerTypes/PayrollSettings/PayrollForm.ts
import { BooleanEditor, DecimalEditor, LookupEditor, StringEditor, IntegerEditor, DateEditor, TextAreaEditor, EnumEditor, PrefixedContext } from "@serenity-is/corelib";
import { EISClass } from "../EmployeeProfile/EISClass";
import { SOCSOClass } from "../EmployeeProfile/SOCSOClass";
import { EPFClass } from "../EmployeeProfile/EPFClass";
import { HRDFClass } from "../EmployeeProfile/HRDFClass";
import { MaritalStatus } from "../EmployeeProfile/MaritalStatus";
import { MoneyClaimApplicationEditor } from "@/HumanResource/MoneyClaimApplication/MoneyClaimApplication/MoneyClaimApplicationEditor";
import { PayslipPaidOneTimeAllowance } from "@/HumanResource/PayrollSettings/PayslipPaidOneTimeAllowance/PayslipPaidOneTimeAllowanceEditor";
import { PayslipDeductedOneTimeDeductions } from "@/HumanResource/PayrollSettings/PayslipDeductedOneTimeDeductions/PayslipDeductedOneTimeDeductionsEditor";
import { PayrollEarningsEditor } from "@/HumanResource/PayrollSettings/PayrollEarnings/PayrollEarningsEditor";
import { PayrollDeductionsEditor } from "@/HumanResource/PayrollSettings/PayrollDeductions/PayrollDeductionsEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface PayrollForm {
    IncentiveSubjectEpf: BooleanEditor;
    IncentiveSubjectEis: BooleanEditor;
    IncentiveSubjectPcb: BooleanEditor;
    IncentiveSubjectSocso: BooleanEditor;
    IncentiveSubjectHrdf: BooleanEditor;
    BonusSubjectEpf: BooleanEditor;
    BonusSubjectEis: BooleanEditor;
    BonusSubjectPcb: BooleanEditor;
    BonusSubjectSocso: BooleanEditor;
    BonusSubjectHrdf: BooleanEditor;
    AnnualizedIncentive: BooleanEditor;
    SeperateIncentive: BooleanEditor;
    AnnualizedBonus: BooleanEditor;
    SeperateBonus: BooleanEditor;
    EmployeeCp38: DecimalEditor;
    EmployeeRowId: LookupEditor;
    EmployeeName: StringEditor;
    PayMonth: IntegerEditor;
    PayYear: IntegerEditor;
    PayDate: DateEditor;
    PayPeriodStart: DateEditor;
    PayPeriodEnd: DateEditor;
    Remarks: TextAreaEditor;
    HrdfWages: DecimalEditor;
    PcbWages: DecimalEditor;
    EisWages: DecimalEditor;
    SocsoWages: DecimalEditor;
    EpfWages: DecimalEditor;
    DaysWorked: DecimalEditor;
    BasicPay: DecimalEditor;
    DailyRate: DecimalEditor;
    HourlyRate: DecimalEditor;
    Age: DecimalEditor;
    BirthDay: DateEditor;
    EisClass: EnumEditor;
    SocsoClass: EnumEditor;
    EpfClass: EnumEditor;
    HrdfClass: EnumEditor;
    TaxClass: IntegerEditor;
    MaritalStatus: EnumEditor;
    WorkingSpouse: BooleanEditor;
    ChildrenUnderEighteen: DecimalEditor;
    ChildrenInUniversity: DecimalEditor;
    DisabledChild: DecimalEditor;
    DisabledChildInUniversity: DecimalEditor;
    Bonus: DecimalEditor;
    Incentive: DecimalEditor;
    EarlyLeavingRate: DecimalEditor;
    LateArrivalRate: DecimalEditor;
    EarlyLeaving: DecimalEditor;
    LateArrival: DecimalEditor;
    FlatOt: DecimalEditor;
    OtOnePointFiveRate: DecimalEditor;
    OtTwoRate: DecimalEditor;
    OtOne: DecimalEditor;
    OtOnePointFive: DecimalEditor;
    OtTwo: DecimalEditor;
    NPLHourlyRate: DecimalEditor;
    NPLDailyRate: DecimalEditor;
    NPLHourly: DecimalEditor;
    NPLDaily: DecimalEditor;
    AbsentDailyRate: DecimalEditor;
    AbsentDaily: DecimalEditor;
    OtSubjectEpf: BooleanEditor;
    OtSubjectEis: BooleanEditor;
    OtSubjectPcb: BooleanEditor;
    OtSubjectSocso: BooleanEditor;
    OtSubjectHrdf: BooleanEditor;
    PaidMoneyClaimingList: MoneyClaimApplicationEditor;
    AllowanceList: PayslipPaidOneTimeAllowance;
    DeductionList: PayslipDeductedOneTimeDeductions;
    PayrollEarnings: PayrollEarningsEditor;
    PayrollDeductions: PayrollDeductionsEditor;
    Deduction: DecimalEditor;
    Earnings: DecimalEditor;
    Nett: DecimalEditor;
    EmployeeEIS: DecimalEditor;
    EmployeePCB: DecimalEditor;
    EmployeeSOCSO: DecimalEditor;
    EmployeeEPF: DecimalEditor;
    EmployerHRDF: DecimalEditor;
    EmployerEPF: DecimalEditor;
    EmployerEIS: DecimalEditor;
    EmployerSOCSO: DecimalEditor;
}

export class PayrollForm extends PrefixedContext {
    static formKey = 'PayrollSettings.Payroll';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!PayrollForm.init)  {
            PayrollForm.init = true;

            var w0 = BooleanEditor;
            var w1 = DecimalEditor;
            var w2 = LookupEditor;
            var w3 = StringEditor;
            var w4 = IntegerEditor;
            var w5 = DateEditor;
            var w6 = TextAreaEditor;
            var w7 = EnumEditor;
            var w8 = MoneyClaimApplicationEditor;
            var w9 = PayslipPaidOneTimeAllowance;
            var w10 = PayslipDeductedOneTimeDeductions;
            var w11 = PayrollEarningsEditor;
            var w12 = PayrollDeductionsEditor;

            initFormType(PayrollForm, [
                'IncentiveSubjectEpf', w0,
                'IncentiveSubjectEis', w0,
                'IncentiveSubjectPcb', w0,
                'IncentiveSubjectSocso', w0,
                'IncentiveSubjectHrdf', w0,
                'BonusSubjectEpf', w0,
                'BonusSubjectEis', w0,
                'BonusSubjectPcb', w0,
                'BonusSubjectSocso', w0,
                'BonusSubjectHrdf', w0,
                'AnnualizedIncentive', w0,
                'SeperateIncentive', w0,
                'AnnualizedBonus', w0,
                'SeperateBonus', w0,
                'EmployeeCp38', w1,
                'EmployeeRowId', w2,
                'EmployeeName', w3,
                'PayMonth', w4,
                'PayYear', w4,
                'PayDate', w5,
                'PayPeriodStart', w5,
                'PayPeriodEnd', w5,
                'Remarks', w6,
                'HrdfWages', w1,
                'PcbWages', w1,
                'EisWages', w1,
                'SocsoWages', w1,
                'EpfWages', w1,
                'DaysWorked', w1,
                'BasicPay', w1,
                'DailyRate', w1,
                'HourlyRate', w1,
                'Age', w1,
                'BirthDay', w5,
                'EisClass', w7,
                'SocsoClass', w7,
                'EpfClass', w7,
                'HrdfClass', w7,
                'TaxClass', w4,
                'MaritalStatus', w7,
                'WorkingSpouse', w0,
                'ChildrenUnderEighteen', w1,
                'ChildrenInUniversity', w1,
                'DisabledChild', w1,
                'DisabledChildInUniversity', w1,
                'Bonus', w1,
                'Incentive', w1,
                'EarlyLeavingRate', w1,
                'LateArrivalRate', w1,
                'EarlyLeaving', w1,
                'LateArrival', w1,
                'FlatOt', w1,
                'OtOnePointFiveRate', w1,
                'OtTwoRate', w1,
                'OtOne', w1,
                'OtOnePointFive', w1,
                'OtTwo', w1,
                'NPLHourlyRate', w1,
                'NPLDailyRate', w1,
                'NPLHourly', w1,
                'NPLDaily', w1,
                'AbsentDailyRate', w1,
                'AbsentDaily', w1,
                'OtSubjectEpf', w0,
                'OtSubjectEis', w0,
                'OtSubjectPcb', w0,
                'OtSubjectSocso', w0,
                'OtSubjectHrdf', w0,
                'PaidMoneyClaimingList', w8,
                'AllowanceList', w9,
                'DeductionList', w10,
                'PayrollEarnings', w11,
                'PayrollDeductions', w12,
                'Deduction', w1,
                'Earnings', w1,
                'Nett', w1,
                'EmployeeEIS', w1,
                'EmployeePCB', w1,
                'EmployeeSOCSO', w1,
                'EmployeeEPF', w1,
                'EmployerHRDF', w1,
                'EmployerEPF', w1,
                'EmployerEIS', w1,
                'EmployerSOCSO', w1
            ]);
        }
    }
}

[EISClass, SOCSOClass, EPFClass, HRDFClass, MaritalStatus]; // referenced types

File: Modules/HumanResource/PayrollSettings/Payroll/PayrollForm.cs
using HRMSoftware.EmployeeProfile;
using HRMSoftware.MoneyClaimApplication;
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;

namespace HRMSoftware.PayrollSettings.Forms;

[FormScript("PayrollSettings.Payroll")]
[BasedOnRow(typeof(PayrollRow), CheckNames = true)]
public class PayrollForm
{
    [HalfWidth]
    public int EmployeeRowId { get; set; }

    [HalfWidth]
    public string EmployeeName { get; set; }
    [OneThirdWidth]
    public int PayMonth { get; set; }
    [OneThirdWidth]
    public int PayYear { get; set; }
    [OneThirdWidth]
    public DateTime PayDate { get; set; }

    [HalfWidth]
    public DateTime PayPeriodStart { get; set; }
    [HalfWidth]
    public DateTime PayPeriodEnd { get; set; }

    [TextAreaEditor(Rows = 3)]
    public string Remarks { get; set; }

    //
    public double DaysWorked { get; set; }
    public double BasicPay { get; set; }
    public double DailyRate { get; set; }
    public double HourlyRate { get; set; }
    public double Age { get; set; }
    public DateTime BirthDay { get; set; }


    public int EisClass { get; set; }
    public int SocsoClass { get; set; }
    public int EpfClass { get; set; }
    public int HrdfClass { get; set; }

    public int TaxClass { get; set; }
    //



    public int MaritalStatus { get; set; }
    public int WorkingSpouse { get; set; }
    public double ChildrenUnderEighteen { get; set; }
    public double ChildrenInUniversity { get; set; }
    public double DisabledChild { get; set; }
    public double DisabledChildInUniversity { get; set; }

    //
    public double EarlyLeavingRate { get; set; }
    public double LateArrivalRate { get; set; }
    public double EarlyLeaving { get; set; }
    public double LateArrival { get; set; }
    public double FlatOt { get; set; }
    public double OtOnePointFiveRate { get; set; }
    public double OtTwoRate { get; set; }

    public double OtOne { get; set; }
    public double OtOnePointFive { get; set; }
    public double OtTwo { get; set; }


    public double NPLHourlyRate { get; set; }
    public double NPLDailyRate { get; set; }
    public double NPLHourly { get; set; }
    public double NPLDaily { get; set; }
    public double AbsentDailyRate { get; set; }

    public double AbsentDaily { get; set; }
    /*
    public double TotalAllowance { get; set; }
    public double TotalDeduction { get; set; }
    public double EpfWage { get; set; }
    public double HrdfWage { get; set; }
    public double SocsoWage { get; set; }
    public double EisWage { get; set; }
    public double TotalTaxableWage { get; set; }
    public double GrossWage { get; set; }
    public double NettWage { get; set; }
    */
    public bool OtSubjectEpf { get; set; }
    public bool OtSubjectEis { get; set; }
    public bool OtSubjectPcb { get; set; }
    public bool OtSubjectSocso { get; set; }
    public bool OtSubjectHrdf { get; set; }

    //


    [IgnoreName, LabelWidth("0"), MoneyClaimApplicationEditor]
    public List<MoneyClaimApplicationRow> PaidMoneyClaimingList { get; set; }

    //
    /*
    public List<int> PaidOtList { get; set; }
    public List<int> DeductedLateArrivalList { get; set; }
    public List<int> DeductedNoPaidLeaveList { get; set; }
    public List<int> DeductedEarlyLeavingList { get; set; }
    */
    [IgnoreName, LabelWidth("0"), PayslipPaidOneTimeAllowance]
    public List<PayslipPaidOneTimeAllowanceRow> AllowanceList { get; set; }
    [IgnoreName, LabelWidth("0"), PayslipDeductedOneTimeDeductions]
    public List<PayslipDeductedOneTimeDeductionsRow> DeductionList { get; set; }

    [IgnoreName, LabelWidth("0"), PayrollEarningsEditor]
    [HideOnInsert, HideOnUpdate]
    public List<PayrollEarningsRow> PayrollEarnings { get; set; }

    [IgnoreName, LabelWidth("0"), PayrollDeductionsEditor]
    [HideOnInsert, HideOnUpdate]
    public List<PayrollDeductionsRow> PayrollDeductions { get; set; }






    public double Deduction { get; set; }
    public double Earnings { get; set; }
    public double Nett { get; set; }
    public double EmployeeEIS { get; set; }
    public double EmployeePCB { get; set; }
    public double EmployeeSOCSO { get; set; }
    public double EmployeeEPF { get; set; }

    public double EmployerHRDF { get; set; }
    public double EmployerEPF { get; set; }
    public double EmployerEIS { get; set; }
    public double EmployerSOCSO { get; set; }

}